scratchai/attacks/attacks/deepfool.py

```diff
@@ -38,47 +38,63 @@ def deepfool(x, net:nn.Module, eta:float=0.02, tnc:int=10, miter:int=50,
   # Set net to eval
   net.eval()
   dev = x.device
+
+  # If x doesn't have a batch dimension add 1
+  # The loop runs twice to ensure that 2D images gets the batch dim
+  # For 2D images, 1st iter adds -> The channel dim
+  # 2nd iter adds -> batch dim
+  for _ in range(2): 
+    if len(x.shape) < 3: x = x.unsqueeze(0)
+  bs = x.shape[0]
+  lbs = list(range(bs))
   
   # x_idt is the initial advesarial image
   x_idt = x.detach().clone().requires_grad_(True)
-  # Get the logits (since x_idt is not perturbed yet, the logits are true logits)
-  logits = net(x_idt).squeeze()
+  # Get the logits (x_idt is not perturbed yet, the logits are true logits)
+  logits = net(x_idt)
   # Get the top tnc classes that the net is most confident about
-  psort = logits.data.cpu().numpy().flatten().argsort()[::-1][:tnc]
+  psort = torch.argsort(logits.data.cpu(), dim=1, descending=True)
+  psort = psort[:, :tnc].data.numpy()
   
-  tlabl = psort[0] # True Label
-  plabl = tlabl    # Pert label
+  tlabl = psort[:, 0] # True Label
+  plabl = tlabl       # Pert label
 
-  x_shape = x_idt.squeeze().shape # [C x H x W]
-  w = np.zeros(x_shape)           # [C x H x W]
-  rt = np.zeros(x_shape)          # [C x H x W]
+  x_shape = x_idt.shape  # [N x C x H x W]
+  w = np.zeros(x_shape)  # [N x C x H x W]
+  rt = np.zeros(x_shape) # [N x C x H x W]
   
   i = 0
-  while plabl == tlabl and i < miter:
+  while (plabl == tlabl).any() and i < miter:
     # Initial Perturbation
-    pert = np.inf
-    logits[tlabl].backward(retain_graph=True)
+    pert = 5e10
+    for ii in range(bs): logits[ii, tlabl[ii]].backward(retain_graph=True)
     ograd = x_idt.grad.data.cpu().numpy().copy()
 
     for c in range(1, tnc):
       zgrad(x_idt)
-      logits[psort[c]].backward(retain_graph=True)
+      for ii in range(bs): logits[ii, psort[ii, c]].backward(retain_graph=True)
       cgrad = x_idt.grad.data.cpu().numpy().copy()
 
       # Get new wk and fk
       wk = cgrad - ograd
-      fk = (logits[c] - logits[tlabl]).item()
```

Comment:
This is a silent bug. The correct implementation should be
`fk = (logits[psort[c]] - logits[tlabl]).item()`
as we need the index of the most closest `c`th class. And instead `logits[c]` will simply take the `c`th index of the logit (or the `c`th class) irrespective of which class is the closest based on the current input. 
**This is fixed in this PR** with
`fk = (logits[lbs, psort[:, c]] - logits[lbs, tlabl]) \`
as can be seen, its updated to take the index from `psort[:, c]`